File: scala/97.scala
object Solution {
    def isInterleave(s1: String, s2: String, s3: String): Boolean = {
        if (s3.length() != s1.length() + s2.length()) return false
        val dp = Array.ofDim[Boolean](s1.length()+1, s2.length()+1)
        dp(0)(0) = true
        for ( i <- 1 to s1.length()) {
            dp(i)(0) = s1(i-1) == s3(i-1) && dp(i-1)(0)
        }
        for ( j <- 1 to s2.length()) {
            dp(0)(j) = s2(j-1) == s3(j-1) && dp(0)(j-1)
        }
        for (i <- 1 to s1.length()) {
            for (j <- 1 to s2.length()) {
                dp(i)(j) = (s1(i-1) == s3(i+j-1) && dp(i-1)(j)) || (s2(j-1) == s3(i+j-1) && dp(i)(j-1))
            }
        }
        dp.last.last
    }
}

File: leetcodeJava/97.java
package leetcodeJava;

class Solution {
    public boolean isInterleave(String s1, String s2, String s3) {
        if (s1.length() + s2.length() != s3.length()) return false;
        if (s1.length() == 0 && s2.length() == 0) return true;
        var dp = new boolean[s1.length()+1][s2.length()+1];
        dp[0][0] = true;
        for (var i = 0; i < s1.length(); i++) {
            if (s1.charAt(i) == s3.charAt(i))
                dp[i+1][0] = dp[i][0];
        }
        for (var j = 0; j < s2.length(); j++) {
            if (s2.charAt(j) == s3.charAt(j))
                dp[0][j+1] = dp[0][j];
        }
        for (var i = 0; i < s1.length(); i++) {
            for (var j = 0; j < s2.length(); j++) {
                if (s2.charAt(j) == s3.charAt(i+j+1)) {
                    if (dp[i+1][j]) dp[i+1][j+1] = true;
                } 
                if (s1.charAt(i) == s3.charAt(i+j+1)) {
                    if (dp[i][j+1]) dp[i+1][j+1] = true;
                }
            }
        }
        return dp[s1.length()][s2.length()];
    }
}
